src/BSON/Javascript.c

```diff
@@ -411,6 +411,34 @@ phongo_create_object_retval php_phongo_javascript_create_object(zend_class_entry
 #endif
 } /* }}} */
 
+static phongo_create_object_retval php_phongo_javascript_clone_object(zval* object TSRMLS_DC) /* {{{ */
+{
+	php_phongo_javascript_t* intern;
+	php_phongo_javascript_t* new_intern;
+	phongo_create_object_retval new_object;
+
+	intern = Z_JAVASCRIPT_OBJ_P(object);
+	new_object = php_phongo_javascript_create_object(Z_OBJCE_P(object) TSRMLS_CC);
+
+#if PHP_VERSION_ID >= 70000
+	new_intern = Z_OBJ_JAVASCRIPT(new_object);
+	zend_objects_clone_members(&new_intern->std, &intern->std TSRMLS_CC);
+#else
+	{
+		zend_object_handle handle = Z_OBJ_HANDLE_P(object);
+
+		new_intern = (php_phongo_javascript_t*) zend_object_store_get_object_by_handle(new_object.handle TSRMLS_CC);
+		zend_objects_clone_members(&new_intern->std, new_object, &intern->std, handle TSRMLS_CC);
+	}
+#endif
+
+	php_phongo_javascript_init(new_intern, intern->code, intern->code_len, NULL TSRMLS_CC);
+	new_intern->scope = bson_new();
+	bson_copy_to(intern->scope, new_intern->scope);
```

Comment:
Documentation for [`bson_copy_to()`](http://mongoc.org/libbson/current/bson_copy_to.html) states that "`dst` must be an uninitialized `bson_t` to avoid leaking memory." If you run the corresponding Javascript clone test with `TEST_PHP_ARGS=-m`, does this come up as a leak?

Reply:
I have to set up proper non-MacOS tooling as valgrind still doesn't work with MacOS 😔

A quick look through other usages of `bson_copy_to` reveals that `bson_destroy` is called beforehand many times. I've added the same here, especially since it does nothing if the BSON is not initialised. Great catch!

Reply:
I can test this locally and get back to you.

Reply:
Definitely a leak here:

```
128 bytes in 1 blocks are definitely lost in loss record 25 of 30
   at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
   by 0x7E9251: __zend_malloc (zend_alloc.c:2829)
   by 0xE58C2E7: php_phongo_malloc (php_phongo.c:2743)
   by 0xE5018B4: bson_malloc (bson-memory.c:70)
   by 0xE4EF2F7: bson_new (bson.c:2039)
   by 0xE5987BD: php_phongo_javascript_clone_object (Javascript.c:440)
   by 0x8D586D: ZEND_CLONE_SPEC_CV_HANDLER (zend_vm_execute.h:33778)
   by 0x911AC6: execute_ex (zend_vm_execute.h:61991)
   by 0x91402D: zend_execute (zend_vm_execute.h:63776)
   by 0x8239AE: zend_execute_scripts (zend.c:1502)
   by 0x787656: php_execute_script (main.c:2590)
   by 0x916CD9: do_cli (php_cli.c:1011)
```

> FWIW, removing the `bson_new` call above leads to a segmentation fault.

This appears to be due to the `BSON_ASSERT(dst)`. I observed the following error when running the clone test with `bson_new` removed:

```
src/libmongoc/src/libbson/src/bson/bson.c:2194 bson_copy_to(): precondition failed: dst
Aborted (core dumped)
```

It appears [`bson_copy_to()`](http://mongoc.org/libbson/current/bson_copy_to.html) expects to receive a pointer to an uninitialized `bson_t` struct. In practice, that would be the address of a stack-allocated `bson_t`, rather than something returned by `bson_new()`. Alternatively, you could `calloc` a block of memory with the size of a `bson_t`, but that's unconventional.

[`bson_copy()`](http://mongoc.org/libbson/current/bson_copy.html) seems more appropriate here:

```c
new_intern->scope = bson_copy(intern->scope);
```

Reply:
Ah, `bson_copy` sounds like it does what I was looking for. I saw `bson_copy_to` being used in a few places and got confused 😕